### PR TITLE
Feat: general wrapper

### PR DESCRIPTION
**Pull request by younesStrittmatter**
# Description

a general wrapper that accepts arguments to field mappings as arguments

# Type of change
- **feat**: A new feature

### CODE REVIEW

File: src/autora/state/wrapper.py [hunk 1]
```diff
@@ -85,3 +88,164 @@ def experimentalist(params):
...
+        >>> from dataclasses import dataclass, field
+        >>> import pandas as pd
+        >>> from typing import List, Optional
+
+        The `State` it operates on needs to have the metadata described in the state module:
+        >>> @dataclass(frozen=True)
+        ... class S(State):
+        ...     conditions: List[int] = field(metadata={"delta": "replace"})
+
+        We define an arbitraty function
```
Comment (benwandrew) [inline src/autora/state/wrapper.py:120]:
```suggestion
        We define an arbitrary function
```

File: src/autora/state/wrapper.py [hunk 2]
```diff
@@ -85,3 +88,164 @@ def experimentalist(params):
...
+        >>> def theorist(x: pd.DataFrame, v: VariableCollection, **kwargs):
+        ...     ivs = [v.name for v in v.independent_variables]
+        ...     dvs = [v.name for v in v.dependent_variables]
+        ...     X, y = x[ivs], x[dvs]
+        ...     new_model = LinearRegression(fit_intercept=True).set_params(**kwargs).fit(X, y)
+        ...     return new_model
+
+        Again, we use the wrapper:
+        >>> theorist_on_state = wrap_to_use_state_general(
+        ...                                         theorist,
```
Comment (benwandrew) [inline src/autora/state/wrapper.py:165]: do we want to include keywords?

```suggestion
        ...                                         f=theorist,
```

File: src/autora/state/wrapper.py [hunk 3]
```diff
@@ -85,3 +88,164 @@ def experimentalist(params):
...
+        ...                                         {'x':'experiment_data', 'v':'variables'},
```
Comment (benwandrew) [inline src/autora/state/wrapper.py:166]:
```suggestion
        ...                                         input_state_mapping={'x':'experiment_data', 'v':'variables'},
```

File: src/autora/state/wrapper.py [hunk 4]
```diff
@@ -85,3 +88,164 @@ def experimentalist(params):
...
+        ...                                          'model')
```
Comment (benwandrew) [inline src/autora/state/wrapper.py:167]:
```suggestion
        ...                                          output_field_name='model')
```

File: src/autora/state/wrapper.py [hunk 5]
```diff
@@ -85,3 +88,164 @@ def experimentalist(params):
...
+        (array([[1.]]), array([2.]))
+
+
+        Other arguments are still supported by the inner function can also be passed
+        (if and only if the inner function allows for and handles `**kwargs` arguments alongside
+        the values from the state).
+        >>> theorist_on_state(t, fit_intercept=False).model.intercept_
+        0.0
+
+        Any parameters not provided by the state can still be provided by to the wrapped function
```
Comment (benwandrew) [inline src/autora/state/wrapper.py:180]:
```suggestion
        Any parameters not provided by the state can still be provided to the wrapped function
```

File: src/autora/state/wrapper.py [hunk 6]
```diff
@@ -85,3 +88,164 @@ def experimentalist(params):
...
+        ...     new_conditions = [c + offset for c in x]
+        ...     return new_conditions
+
+        >>> experimentalist_on_state_ = wrap_to_use_state_general(
+        ...                                                 f=experimentalist,
+        ...                                                 input_state_mapping={'x':'conditions'},
+        ...                                                 output_field_name='conditions')
+
+
+        ... then it need to be passed.
```
Comment (benwandrew) [inline src/autora/state/wrapper.py:191]: need this? not quite sure how to read as a continuation of previous sentence (which seems comprehensive)

**Review by benwandrew — COMMENTED** [review]
this is super cool! i hope John Ousterhout* would like the direction we're moving in ;)

few suggestions and fixed a pre-commit error, but looking very good!

one general question for my own understanding (which may be better for me to address in office hours): i'm confused about the relationship between the decorator/wrapper `wrap_to_use_state` in `delta.py` and this new `wrap_to_use_state_general` in `wrapper.py`. is the latter not ultimately replacing the former?

*reference to author of a book @hollandjg shared a while ago: https://milkov.tech/assets/psd.pdf

**Comment by hollandjg** [comment]
I like the look of this!

Would it be sensible to split up the wrapper into two – one which does the input name mapping, and one which does the output wrapping? Then we could use the two independently – perhaps your inner function needs to return two values (say `conditions` and `experiment_js_code`) and returns those as a `Result`/`Delta` object, but you still want a mapping. Conversely, perhaps you have a function which uses entirely standard naming for the variables, but you want to wrap the outputs.

File: src/autora/state/wrapper.py [hunk 7]
```diff
@@ -85,3 +88,166 @@ def experimentalist(params):
         return Delta(conditions=conditions_)
 
     return experimentalist
+
+
+class WrapperInputError(Exception):
+    pass
+
+
+def wrap_to_use_state_general(
```
Comment (musslick) [inline src/autora/state/wrapper.py:97]: Could the name be shorted, e.g., ``wrap_to_state``? I imagine people will be using this a lot, so we might want to give it a short name.

**Review by musslick — CHANGES_REQUESTED** [review]
Looks great; just got a minor comment for a name change.

**Comment by hollandjg** [comment]
hey @younesStrittmatter , do you think you could refactor this so that there's a wrapper function which just does the input field-name mapping? We could include that as an option in the `on_state` function from #33, but it needs to be an independent wrapper function first so that we can test it really extensively on its own.

**Comment by hollandjg** [comment]
I think this is covered by the newest version of the state object – closing this now.

**Pull request closed without merging**